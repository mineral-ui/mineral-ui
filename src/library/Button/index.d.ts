import {
  ButtonHTMLAttributes,
  DetailedHTMLProps,
  ElementType,
  ReactElement
} from 'react';

export declare namespace Button {
  export interface Props
    extends DetailedHTMLProps<
      ButtonHTMLAttributes<HTMLButtonElement>,
      HTMLButtonElement
    > {
    as?: ElementType;
    circular?: boolean;
    css?: { [key: string]: any };
    fullWidth?: boolean;
    href?: string;
    iconEnd?: ReactElement;
    iconStart?: ReactElement;
    justifyContent?: string;
    minimal?: boolean;
    primary?: boolean;
    size?: string;
    variant?: string;
    width?: string;
  }
}

export declare function Button(props: Button.Props): ReactElement;
export default Button;
