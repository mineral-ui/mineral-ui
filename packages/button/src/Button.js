/**
 * Copyright 2017 CA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */
import React from 'react';
import { createStyledComponent } from '@mineral-ui/style-utils';

type Props = {|
  children?: MnrlReactNode,
  className?: string,
  disabled?: boolean,
  onPress: Function,
  size?: 'small' | 'regular' | 'big',
  type?: 'button' | 'submit',
  variant?: 'danger' | 'minimal' | 'primary' | 'regular' | 'success' | 'warning'
|};

const Root = createStyledComponent('button', (props, theme) => ({
  // TODO: global reset
  boxSizing: 'border-box',
  fontFamily: `${theme.fontFamily}, ${theme.fontFamily_system}`,
  lineHeight: theme.lineHeight,
  outline: 0,

  backgroundColor: (() => {
    if (props.disabled && props.variant !== 'minimal') {
      return theme.color_gray_30;
    } else {
      return {
        regular: theme.color_gray_10,
        minimal: 'transparent',
        primary: theme.color_background_primary,
        danger: theme.color_background_danger,
        success: theme.color_background_success,
        warning: theme.color_background_warning
      }[props.variant];
    }
  })(),
  borderColor: props.variant === 'regular' && !props.disabled
    ? theme.color_border
    : 'transparent',
  borderRadius: theme.borderRadius_1,
  borderStyle: 'solid',
  borderWidth: '1px',
  boxShadow: props.variant !== 'minimal' &&
    props.variant !== 'regular' &&
    !props.disabled
    ? theme.shadow_1
    : null,
  color: (() => {
    if (props.disabled) {
      return theme.color_gray_50;
    } else {
      return {
        regular: theme.color_gray_80,
        minimal: theme.color_link,
        primary: theme.color_text_onPrimary,
        danger: theme.color_text_onPrimary,
        success: theme.color_text_onPrimary,
        warning: theme.color_text_onPrimary
      }[props.variant];
    }
  })(),
  fontSize: props.size === 'small' ? '0.75rem' : theme.fontSize_ui,
  fontWeight: theme.fontWeight_bold,
  height: theme[`size_${props.size}`],
  paddingLeft: props.size === 'big'
    ? theme.spacing_double
    : theme.spacing_single,
  paddingRight: props.size === 'big'
    ? theme.spacing_double
    : theme.spacing_single,

  '&:hover': {
    backgroundColor: (() => {
      if (!props.disabled) {
        return {
          regular: theme.color_gray_20,
          minimal: theme.color_gray_10,
          primary: theme.color_background_hover_primary,
          danger: theme.color_background_hover_danger,
          success: theme.color_background_hover_success,
          warning: theme.color_background_hover_warning
        }[props.variant];
      }
    })()
  },

  '&:focus': {
    borderColor: (() => {
      return {
        regular: theme.color_border_focus,
        minimal: theme.color_border_focus,
        primary: theme.color_border_focus_primary,
        danger: theme.color_border_focus_danger,
        success: theme.color_border_focus_success,
        warning: theme.color_border_focus_warning
      }[props.variant];
    })()
  },

  '&:active': {
    backgroundColor: (() => {
      return {
        regular: theme.color_theme_10,
        minimal: theme.color_theme_10,
        primary: theme.color_background_active_primary,
        danger: theme.color_background_active_danger,
        success: theme.color_background_active_success,
        warning: theme.color_background_active_warning
      }[props.variant];
    })(),
    borderColor: props.variant === 'regular'
      ? theme.color_border_active
      : 'transparent'
  }
}));

export default function Button({
  children,
  className,
  disabled,
  onPress,
  size = 'regular',
  type = 'button',
  variant = 'regular'
}: Props) {
  const rootProps = {
    className,
    disabled,
    onClick: onPress,
    size,
    type,
    variant
  };

  return <Root {...rootProps}>{children}</Root>;
}
