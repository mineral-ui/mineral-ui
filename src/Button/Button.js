/**
 * Copyright 2017 CA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */
import React, { cloneElement } from 'react';
import { ellipsis } from 'polished';
import { createStyledComponent, pxToEm, getNormalizedValue } from '../styles';

type Props = {
  /** Rendered content of the component */
  children?: React$Node,
  /** Displays a circular Button */
  circular?: boolean,
  /** Disables the Button */
  disabled?: boolean,
  /** Element to be used as the root node - e.g. `a` can be used to create a link that is styled like a Button */
  element?: $FlowFixMe, // Should allow string | React class
  /** Stretch Button to fill its container */
  fullWidth?: boolean,
  /** Icon that goes after the children*/
  iconEnd?: React$Element<*>,
  /** Icon that goes before the children */
  iconStart?: React$Element<*>,
  /** Display a minimal Button */
  minimal?: boolean,
  /** Called with the click event */
  onClick?: (event: SyntheticEvent<>) => void,
  /** Display a primary Button */
  primary?: boolean,
  /** Available sizes */
  size?: 'small' | 'medium' | 'large' | 'jumbo',
  /** Available types */
  type?: string,
  /** Available variants */
  variant?: 'regular' | 'danger' | 'success' | 'warning'
};

export const componentTheme = (baseTheme: Object) => ({
  Button_backgroundColor: baseTheme.color_gray_20,
  Button_backgroundColor_active: baseTheme.color_gray_30,
  Button_backgroundColor_focus: baseTheme.color_gray_20,
  Button_backgroundColor_hover: baseTheme.color_gray_10,
  Button_backgroundColor_minimal_active: baseTheme.color_gray_20,
  Button_backgroundColor_minimal_hover: baseTheme.color_gray_10,
  Button_backgroundColor_primary: baseTheme.color_theme_60,
  Button_backgroundColor_primary_active: baseTheme.color_theme_70,
  Button_backgroundColor_primary_focus: baseTheme.color_theme_60,
  Button_backgroundColor_primary_hover: baseTheme.color_theme_50,
  Button_borderColor: baseTheme.borderColor,
  Button_borderColor_focus: baseTheme.color_white,
  Button_borderRadius: baseTheme.borderRadius_1,
  Button_borderWidth: 1, // px
  Button_boxShadow_focus: `0 0 0 1px ${baseTheme.borderColor_focus}`,
  Button_color_text: baseTheme.color_gray_100,
  Button_color_text_minimal: baseTheme.color_text_primary,
  Button_color_text_primary: baseTheme.color_text_onprimary,
  Button_fontWeight: baseTheme.fontWeight_semiBold,
  Button_paddingHorizontal: baseTheme.space_inset_sm,
  Button_paddingIconOnly_small: pxToEm(3),
  Button_paddingIconOnly_medium: pxToEm(7),
  Button_paddingIconOnly_large: pxToEm(7),
  Button_paddingIconOnly_jumbo: pxToEm(13),
  Button_height_small: baseTheme.size_small,
  Button_height_medium: baseTheme.size_medium,
  Button_height_large: baseTheme.size_large,
  Button_height_jumbo: baseTheme.size_jumbo,

  ButtonContent_fontSize: baseTheme.fontSize_ui,
  ButtonContent_fontSize_small: pxToEm(12),

  ButtonIcon_margin: baseTheme.space_inset_sm,

  ...baseTheme
});

function chooseColor({ disabled, primary, minimal }: Props, theme) {
  if (disabled) {
    return theme.color_text_disabled;
  } else if (primary) {
    return theme.Button_color_text_primary;
  } else if (minimal) {
    return theme.Button_color_text_minimal;
  } else {
    return theme.Button_color_text;
  }
}

const styles = {
  button: props => {
    let theme = componentTheme(props.theme);
    const {
      circular,
      disabled,
      fullWidth,
      minimal,
      primary,
      size,
      text,
      variant
    } = props;

    if (variant !== 'regular') {
      // prettier-ignore
      theme = {
        ...theme,
        Button_backgroundColor_primary: theme[`backgroundColor_${variant}`],
        Button_backgroundColor_primary_active: theme[`backgroundColor_${variant}_active`],
        Button_backgroundColor_primary_focus: theme[`backgroundColor_${variant}_focus`],
        Button_backgroundColor_primary_hover: theme[`backgroundColor_${variant}_hover`],
        Button_boxShadow_focus: `0 0 0 1px ${theme[`borderColor_${variant}_focus`]}`,
        Button_color_text: theme[`color_text_${variant}`],
        Button_color_text_minimal: theme[`color_text_${variant}`],
        Button_color_text_primary: theme[`color_text_on${variant}`]
      };
    }

    const color = chooseColor(props, theme);
    return {
      backgroundColor: (() => {
        if (disabled && !minimal) {
          return theme.color_gray_30;
        } else if (primary) {
          return theme.Button_backgroundColor_primary;
        } else if (minimal) {
          return 'transparent';
        } else {
          return theme.Button_backgroundColor;
        }
      })(),
      borderColor:
        disabled || primary || minimal
          ? 'transparent'
          : theme.Button_borderColor,
      borderRadius: circular
        ? `${parseFloat(theme[`Button_height_${size}`]) / 2}em`
        : theme.Button_borderRadius,
      borderStyle: 'solid',
      borderWidth: `${theme.Button_borderWidth}px`,
      color,
      cursor: disabled ? 'default' : 'pointer',
      display: 'inline-block',
      fontWeight: theme.Button_fontWeight,
      height: theme[`Button_height_${size}`],
      // if the user puts in a small icon in a large button
      // we want to force the button to be round/square
      // (really just pertinent on icon-only buttons)
      minWidth: theme[`Button_height_${size}`],
      padding:
        text === undefined
          ? theme[`Button_paddingIconOnly_${size}`]
          : `0 ${theme.Button_paddingHorizontal}`,
      textDecoration: 'none',
      verticalAlign: 'middle',
      width: fullWidth && '100%',
      '&:focus': {
        backgroundColor: (() => {
          if (primary) {
            return theme.Button_backgroundColor_primary_focus;
          } else if (minimal) {
            return theme.Button_backgroundColor_minimal_focus;
          } else {
            return theme.Button_backgroundColor_focus;
          }
        })(),
        borderColor: theme.Button_borderColor_focus,
        boxShadow: theme.Button_boxShadow_focus,
        color,
        textDecoration: 'none'
      },
      '&:hover': {
        backgroundColor: (() => {
          if (!disabled) {
            if (primary) {
              return theme.Button_backgroundColor_primary_hover;
            } else if (minimal) {
              return theme.Button_backgroundColor_minimal_hover;
            } else {
              return theme.Button_backgroundColor_hover;
            }
          }
        })(),
        color,
        textDecoration: 'none'
      },
      // `:active` must be last, to follow LVHFA order:
      // https://developer.mozilla.org/en-US/docs/Web/CSS/:active
      '&:active': {
        backgroundColor: (() => {
          if (!disabled) {
            if (primary) {
              return theme.Button_backgroundColor_primary_active;
            } else if (minimal) {
              return theme.Button_backgroundColor_minimal_active;
            } else {
              return theme.Button_backgroundColor_active;
            }
          }
        })(),
        color
      },
      '&::-moz-focus-inner': { border: 0 },

      '& [role="icon"]': {
        boxSizing: 'content-box',
        fill:
          disabled || primary || minimal || variant !== 'regular'
            ? 'currentColor'
            : theme.Button_backgroundColor_primary,
        display: 'block',

        '&:first-child': {
          marginLeft:
            theme.direction === 'rtl' ? theme.ButtonIcon_margin : null,
          marginRight:
            theme.direction === 'ltr' ? theme.ButtonIcon_margin : null
        },

        '&:last-child': {
          marginLeft:
            theme.direction === 'ltr' ? theme.ButtonIcon_margin : null,
          marginRight:
            theme.direction === 'rtl' ? theme.ButtonIcon_margin : null
        },

        '&:only-child': {
          margin: 0
        }
      }
    };
  },
  content: props => {
    const theme = componentTheme(props.theme);
    const { size } = props;

    let paddings;

    const fontSize =
      size === 'small'
        ? theme.ButtonContent_fontSize_small
        : theme.ButtonContent_fontSize;

    if (size === undefined || size === 'large' || size === 'jumbo') {
      const padding = getNormalizedValue(
        theme.Button_paddingHorizontal,
        fontSize
      );
      paddings = {
        '&:first-child': {
          paddingLeft: theme.direction === 'ltr' ? padding : null,
          paddingRight: theme.direction === 'rtl' ? padding : null
        },

        '&:last-child': {
          paddingLeft: theme.direction === 'rtl' ? padding : null,
          paddingRight: theme.direction === 'ltr' ? padding : null
        }
      };
    }

    return {
      ...ellipsis('100%'),

      display: 'block',
      fontSize,
      lineHeight: getNormalizedValue(theme[`Button_height_${size}`], fontSize),
      ...paddings
    };
  },
  inner: {
    alignItems: 'center',
    display: 'inline-flex',
    justifyContent: 'center',
    maxHeight: '100%',
    width: '100%'
  }
};

const Content = createStyledComponent('span', styles.content);
const Inner = createStyledComponent('span', styles.inner);

function isTypeButton(type: ?string) {
  return ['button', 'submit', 'reset'].indexOf(type) !== -1;
}

function filterProps({ element, type }: Props) {
  // When element is a component, e.g. ReactRouterLink,
  // these are not filtered automatically by rootEl
  const invalidComponentProps = ['primary', 'text', 'variant'];
  const invalidLinkProps =
    element === 'a' && isTypeButton(type) ? ['type'] : [];
  return Array.prototype.concat(invalidComponentProps, invalidLinkProps);
}

/**
 * The Button component represents a clickable button.
 * Buttons draw attention to actions that can be performed in your app.
 * Buttons are used to trigger any sort of event.
 *
 * Always use the appropriate variant for your button, and write a clear, concise label to effectively communicate intention.
 */
export default function Button({
  children,
  element = 'button',
  iconStart,
  iconEnd,
  size = 'large',
  type = 'button',
  variant = 'regular',
  ...restProps
}: Props) {
  const rootProps = {
    size,
    text: children,
    type,
    variant,
    ...restProps
  };

  const Root = createStyledComponent(element, styles.button, {
    filterProps: filterProps({ element, type }),
    includeStyleReset: true,
    rootEl: element
  });

  const iconSize = {
    small: 'medium',
    medium: 'medium',
    large: pxToEm(24),
    jumbo: pxToEm(24)
  };
  const startIcon = iconStart
    ? cloneElement(iconStart, { size: iconSize[size], key: 'iconStart' })
    : null;
  const endIcon = iconEnd
    ? cloneElement(iconEnd, { size: iconSize[size], key: 'iconEnd' })
    : null;

  return (
    <Root {...rootProps}>
      <Inner>
        {startIcon}
        {children && <Content size={size}>{children}</Content>}
        {endIcon}
      </Inner>
    </Root>
  );
}
