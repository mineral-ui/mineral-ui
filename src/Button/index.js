/* @flow */
export { default } from './Button';
