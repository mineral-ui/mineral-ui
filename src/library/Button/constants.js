/* @flow */
import { pxToEm } from '../styles';

export const ICON_SIZE = {
  small: 'medium',
  medium: pxToEm(20), // TargetX Custom Style
  large: pxToEm(24),
  jumbo: pxToEm(24)
};

export const SIZE = {
  small: 'small',
  medium: 'medium',
  large: 'large',
  jumbo: 'jumbo'
};

export const VARIANT = {
  grayscale: 'grayscale', // TargetX Custom Variant

  danger: 'danger',
  success: 'success',
  warning: 'warning'
};
