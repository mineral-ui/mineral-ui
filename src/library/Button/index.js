/* @flow */
export { default } from './Button';

export * from './themes';
