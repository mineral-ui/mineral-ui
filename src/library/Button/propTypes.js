/* @flow */
import { bool, element, func, node, oneOf, string } from 'prop-types';
import { SIZE, VARIANT } from './constants';

export const buttonPropTypes = {
  width: string,
  justifyContent: string,

  children: node,
  circular: bool,
  disabled: bool,
  fullWidth: bool,
  iconEnd: element,
  iconStart: element,
  minimal: bool,
  onClick: func,
  primary: bool,
  size: oneOf(Object.keys(SIZE)),
  type: string,
  variant: oneOf(Object.keys(VARIANT))
};
