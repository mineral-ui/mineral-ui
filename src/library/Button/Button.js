/* @flow */
import React, { Component, cloneElement } from 'react';
import memoizeOne from 'memoize-one';
import { createButtonRootNode, Content, Inner } from './styled';
import { ICON_SIZE, SIZE } from './constants';

import { buttonPropTypes } from './propTypes';
import type { ButtonDefaultProps, ButtonProps } from './types';

export default class Button extends Component<ButtonProps> {
  static displayName = 'Button';

  static defaultProps: ButtonDefaultProps = {
    element: 'button',
    size: SIZE.large,
    type: 'button'
  };

  static propTypes = buttonPropTypes;

  // Must be an instance method to avoid affecting other instances memoized keys
  getRootNode = memoizeOne(
    createButtonRootNode,
    (nextProps: ButtonProps, prevProps: ButtonProps) =>
      nextProps.element === prevProps.element
  );

  render() {
    const {
      children,
      disabled,
      iconStart,
      iconEnd,
      size = Button.defaultProps.size,
      type = Button.defaultProps.type,
      variant,
      ...restProps
    } = this.props;

    const rootProps = {
      disabled,
      size,
      tabIndex: disabled ? -1 : undefined,
      text: children,
      type,
      variant,
      ...restProps
    };

    const Root = this.getRootNode(this.props, Button.defaultProps);

    const startIcon = iconStart
      ? cloneElement(iconStart, { size: ICON_SIZE[size], key: 'iconStart' })
      : null;
    const endIcon = iconEnd
      ? cloneElement(iconEnd, { size: ICON_SIZE[size], key: 'iconEnd' })
      : null;

    return (
      <Root {...rootProps}>
        <Inner>
          {startIcon}
          {children && <Content size={size}>{children}</Content>}
          {endIcon}
        </Inner>
      </Root>
    );
  }
}
