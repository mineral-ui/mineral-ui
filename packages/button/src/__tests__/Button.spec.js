/**
 * Copyright 2017 CA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow */
import React from 'react';
import { mount, shallow } from 'enzyme';
import Button from '../Button';

function shallowButton() {
  return shallow(<Button onPress={() => {}} />);
}

function mountButton() {
  return mount(<Button onPress={() => {}} />);
}

describe('Button', () => {
  it('renders', () => {
    const button = shallowButton();

    expect(button.exists()).toEqual(true);
  });

  it('renders correctly', () => {
    const button = mountButton();

    expect(button).toMatchSnapshotWithGlamor();
  });
});
