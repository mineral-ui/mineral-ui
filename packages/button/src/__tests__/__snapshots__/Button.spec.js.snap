// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly 1`] = `
<Button>
  <glamorous(button)
    size="regular"
  >
    <button
      className="css-nil"
    />
  </glamorous(button)>
</Button>
`;
